pageserver/src/tenant/remote_timeline_client.rs

```diff
@@ -642,9 +642,86 @@ impl RemoteTimelineClient {
         self.launch_queued_tasks(upload_queue);
     }
 
+    pub(crate) async fn schedule_reparenting_and_wait(
+        self: &Arc<Self>,
+        new_parent: &TimelineId,
+    ) -> anyhow::Result<()> {
+        // FIXME: because of how Timeline::schedule_uploads works when called from layer flushing
+        // and reads the in-memory part we cannot do the detaching like this
+        let receiver = {
+            let mut guard = self.upload_queue.lock().unwrap();
+            let upload_queue = guard.initialized_mut()?;
+
+            upload_queue.latest_metadata.reparent(new_parent);
+
+            self.schedule_index_upload(upload_queue, upload_queue.latest_metadata.clone());
+
+            self.schedule_barrier0(upload_queue)
+        };
+
+        Self::wait_completion0(receiver).await
+    }
+
+    /// Schedule an index part update as the last step of timeline detach.
+    ///
+    /// Metadata of adoption is added to the next index part recording the path for WAL based
+    /// disaster recovery.
     ///
-    /// Launch an upload operation in the background.
+    /// The returned future waits for the upload to complete, which means that remote storage is
+    /// now in sync, and the timeline can be separated from its ancestor. The next time the
+    /// timeline is loaded from remote storage, it will have no ancestor.
     ///
+    /// Cancellation-safety: cancelling or not waiting for the returned future will not cancel the
+    /// upload.
+    pub(crate) async fn schedule_detaching_from_ancestor_and_wait(
+        self: &Arc<Self>,
+        adopted: (TimelineId, Lsn),
+    ) -> anyhow::Result<()> {
+        let receiver = {
+            let mut guard = self.upload_queue.lock().unwrap();
+            let upload_queue = guard.initialized_mut()?;
+
+            // TODO: update lineage in index_part once we have it
+            upload_queue
+                .latest_metadata
+                .detach_from_ancestor(&adopted.0, &adopted.1);
```

Comment:
issue link or it ~~didn't~~ won't happen :)

Reply:
I have a list of follow-ups on the epic, lineage was mentioned earlier in https://github.com/neondatabase/neon/pull/7456#discussion_r1587833332.